design/automatic-network-configuration.md

```diff
@@ -0,0 +1,801 @@
+    4. Update the port CR (status) (Configuring step).
+    5. Port is in Active step.
+9. BMO detects that the status.state of the Port CR corresponding to all BMH
+network cards is empty (without config in the specs) or Active,
+and then continues to provision BMH.
+
+#### Update Workflow
+
+1. The Switch controller detects that Port.spec.configurationRef changed,
+then carry out the corresponding processing:
+    1. Verify the user configuration to match the switch configuration
+    (allowed vlans etc.) (Validating step).
+    2. Credentials verification (Validating step).
+    3. Apply the configuration to the port on the switch (Configuring step).
+    4. Update the port CR (status) (Configuring step).
+    5. Port is in Active step.
+
+#### Deprovisioning workflow
+
+1. CAPI / User set the deletion timestamp on the `Metal3Machine` CR.
+2. CAPM3 calls the `deconfigureNetwork()` to find all the `SwitchPort` CR
+corresponding to the BMO network cards, and clear their
+`spec.configurationRef`.
+    1. CAPM3 fetches the BMH.
+    2. CAPM3 removes the configurationRef field value
+3. BMO deconfigures the ports.
+    1. BMO fetches all SwitchPorts referenced by the BMH.
+    2. BMO removes any SwitchPortConfig that might be set on the port.
+4. SwitchPort controller starts to deconfigure the network by calling
+`device.DeConfigurePort()`. Return the port to a `cleaning` state.
+    1. Verify the switch credentials (Cleaning Step).
+    2. Configure the port on the switch (Cleaning Step).
+    3. If timeout reached, goes to error state.
+    4. Update the switchPort CR.
+    5. SwitchPort in Idle state.
+5. BMO detects that the status.state of the Port CR corresponding to all
+BMH network cards are idle, and then continues to deprovision BMH.
+If any of the ports are in error state -> goes to error state.
+
+#### Deletion workflow
+
+1. User sets deletion timestamp on all resources (that automatically triggers
+the deletion of the Provider Switch because of OwnerRefs).
+2. Switch deletion is blocked until all SwitchPorts are deleted.
+3. Wait for switch port to be idle.
+4. If switchPort is idle, then remove the finalizer.
+5. Once no ports are left for the Switch, then remove the finalizer on
+the ProviderSwitch and on the Switch.
+
+### Changes to Current API
+
+#### Metal3Machine CRD
+
+```yaml
+spec:
+  # Specify the network configuration that metal3Machine needs to use
+  networkConfiguration:
+  # no smartNic
+  - ConfigurationRef:
+    name: nc1
+    kind: SwitchPortConfiguration
+    namespace: default
+    # Network card performance required for this network configuration
+    nicHint:
+      name: eth0
+      smartNIC: false
+  # with smartNic
+  - ConfigurationRef:
+    name: nc1
+    kind: SmartNicConfiguration
+    namespace: default
+    nicHint:
+      name: eth1
+      smartNIC: true
+```
+
+#### BareMetalHost
+
+Add the `ports` field to `.spec` to indicate the connection information and
```

Comment:
MAC addresses may not be known until inspection, how will it be handled here?

Reply:
In our design, the person who created the BMH needs to know the MAC address